pyjob/sge.py

```diff
@@ -65,27 +66,62 @@ def info(self):
                     data[kv[0]] = kv[1]
         return data
 
-    @property
-    def available_environments(self):
-        """:obj:`~pyjob.sge.SunGridEngineTask` available set of environments information"""
-        try:
-            stdout = cexec(["qconf", "-spl"], permit_nonzero=True)
-        except PyJobExecutableNotFoundError:
-            return []
-        environments = []
+    @staticmethod
+    def sge_avail_config(param):
+        """Get the set of available configurations for a given SGE parameter
+
+        Parameters
+        ----------
+        param : str
+            The parameter to be tested
+
+        Returns
+        -------
+        set
+            A set with the available configurations for the parameter of interest
+
+        Raises
+        ------
+        :exc:`ValueError`
+           Invalid parameter. Supported parameters are 'environment' or 'queue'
+
+        """
+
+        if param == 'environment':
+            cmd = ["qconf", 'spl']
+        elif param == 'queue':
+            cmd = ["qconf", 'sql']
+        else:
+            raise ValueError('Invalid parameter {}. Available parameters are "environment" or "queue".'.format(param))
+
+        stdout = cexec(cmd, permit_nonzero=True)
+        config = []
         for line in stdout.splitlines():
             line = line.split()
             if len(line) > 1:
-                return environments
+                return config
             else:
-                environments.append(line[0].encode('utf-8'))
-        return environments
+                config.append(line[0].encode('utf-8'))
+        return set(config)
```

Comment:
I don't think this is a good implementation.

1/ instead of hard-coded strings, use an `Enum` that defines `environment` and `queue`.
```
from enum import Enum
class SGEConfigParameter(Enum):
    ENVIRONMENT = 1
    QUEUE = 2
```
2/ Calling this *N*-times results in additional subprocess calls despite no change to the environment. how about we cache the output for subsequent calls?
```
_sge_avail_configs_by_env = {}
@classmethod
def get_sge_avail_configs(cls, env):
    if env in cls._sge_avail_configs_by_env: 
        return cls._sge_avail_configs_by_env[env]

    if SGEConfigParameter(env) == SGEConfigParameter.ENVIRONMENT:
        cmd = ['qconf', 'spl']
    elif SGEConfigParameter(env) == SGEConfigParameter.QUEUE:
        cmd = ["qconf", 'sql']
    else:
        raise ValueError('<message>')

    [...]

    cls._sge_avail_configs_by_env[env] = set(config)
    return cls._sge_avail_configs_by_env[env]
```

Reply:
Thanks, I think this is a much cleaner implementation.